File: giphy.py
import time
import giphy_client
from giphy_client.rest import ApiException
import os

# create an instance of the API class
api_instance = giphy_client.DefaultApi()
api_key = os.getenv('GIPHY_TOKEN') # str | Giphy API Key.
rating = 'g' # str | Filters results by specified rating. (optional)
fmt = 'json' # str | Used to indicate the expected response format. Default is Json. (optional) (default to json)

def random_gif(search):
  try: 
    # Search Endpoint
    api_response = api_instance.gifs_random_get(api_key, tag=search, rating=rating, fmt=fmt)
    return api_response.data.image_url
  except ApiException as e:
    return 'http://gph.is/2efdN3V'

File: src/giphy.rs
//! Port of `giphy.py`. Uses the Giphy "random" REST endpoint directly (the Python
//! `giphy_client` library does the same under the hood). Not called by the live
//! bot, but reproduced faithfully including the fallback URL on any error.

const FALLBACK: &str = "http://gph.is/2efdN3V";

pub async fn random_gif(search: &str) -> String {
    let api_key = std::env::var("GIPHY_TOKEN").unwrap_or_default();
    let url = "https://api.giphy.com/v1/gifs/random";
    let result: Result<String, ()> = async {
        let client = reqwest::Client::new();
        let resp = client
            .get(url)
            .query(&[
                ("api_key", api_key.as_str()),
                ("tag", search),
                ("rating", "g"),
                ("fmt", "json"),
            ])
            .send()
            .await
            .map_err(|_| ())?;
        let json: serde_json::Value = resp.json().await.map_err(|_| ())?;
        json.get("data")
            .and_then(|d| d.get("image_url"))
            .and_then(|u| u.as_str())
            .map(|s| s.to_string())
            .ok_or(())
    }
    .await;

    result.unwrap_or_else(|_| FALLBACK.to_string())
}
